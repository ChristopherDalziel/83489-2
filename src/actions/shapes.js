import { v4 as uuidv4 } from "uuid";

export const addShape = ({
  sideOne = "",
  sideTwo = "",
  sideThree = "",
  sideFour = "",
} = {}) => ({
  type: "ADD_SHAPE",
  shape: {
    id: uuidv4(),
    sideOne,
    sideTwo,
    sideThree,
    sideFour,
  },
});
