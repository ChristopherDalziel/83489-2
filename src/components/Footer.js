import React from "react";
import styled from "@emotion/styled";

const FooterDiv = styled.footer`
  margin-top: auto;
  height: 6vh;
  background-color: #e63846;
  border-radius: 50px 0px 0px 0px;
  display: flex;
  align-items: flex-end;
  justify-content: flex-end;
  padding: 2vh;
  font-size: 50px;
`;

const Footer = () => {
  return (
    <FooterDiv>
      <a
        target="_blank"
        href="https://github.com/ChristopherDalziel/83489-2"
        rel="noopener noreferrer"
        class="fab fa-github-square socials"
        id="footerLink"
      >
        <span>Link to projects GitHub repository</span>
      </a>
    </FooterDiv>
  );
};

export default Footer;
