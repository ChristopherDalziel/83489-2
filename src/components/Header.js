import React from "react";
import styled from "@emotion/styled";
import logo from "../assets/logo.png";

const HeaderDiv = styled.header`
  padding: 0;
  margin: 0;
  background-color: #e63846;
  height: 11vh;
  border-radius: 0px 0px 50px 0px;
  box-shadow: 17px 14px 17px -11px rgba(0, 0, 0, 0.2);
  @media (max-width: 815px) {
    border-radius: 0px;
    font-size: 25px;
    height: 10vh;
  }
`;

const Title = styled.h1`
  margin: 0 0 5px 0;
  padding: 0px 10px 0px 10px;
  width: 30vw;
  @media (max-width: 815px) {
    margin-bottom: 5px;
    font-size: 25px;
    width: 75vw;
  }
`;

const SubTitle = styled.h2`
  margin: 0;
  font-size: 1rem;
  width: 20.75vw;
  background-color: black;
  padding: 5px 10px 5px 10px;
  border-radius: 0px 50px 50px 0px;
  @media (max-width: 815px) {
    padding: 5px
    border-radius: 0px;
    font-size: 12px;
    width: 100vw;
  }
`;

const LogoSmall = styled.img`
  height: 90px;
  float: right;
  transition: transform 0.2s;
  :hover {
    -ms-transform: scale(0.9);
    -webkit-transform: scale(0.9);
    transform: scale(0.9);
  }
  @media (max-width: 815px) {
    visibility: hidden;
  }
`;

const Header = (props) => {
  return (
    <HeaderDiv>
      <Title>
        {props.title}
        <LogoSmall src={logo} alt="Site logo" />
      </Title>

      {props.subTitle && <SubTitle>{props.subTitle}</SubTitle>}
    </HeaderDiv>
  );
};

Header.defaultProps = {
  title: "Shape Sorter Online",
  subTitle: "Leave your shape sorting to a computer",
};

export default Header;
