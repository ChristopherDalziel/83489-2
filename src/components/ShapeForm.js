import React from "react";
import styled from "@emotion/styled";
import { connect } from "react-redux";
import {
  sortByTriangles,
  sortByCircles,
  sortByQuadrilaterals,
} from "../actions/filters";

const FormContainer = styled.div`
  padding: 20px 10px 0px 10px;
  color: black;
  @media (max-width: 815px) {
    width: 50vw;
  }
`;

const Instructions = styled.p`
  width: 50vw;
`;

export class ShapeForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      sideOne: props.shape ? props.shape.sideOne : "",
      sideTwo: props.shape ? props.shape.sideTwo : "",
      sideThree: props.shape ? props.shape.sideThree : "",
      sideFour: props.shape ? props.shape.sideFour : "",
      error: "",
    };
  }

  onSideOneChange = (e) => {
    const sideOne = e.target.value;
    if (!sideOne || sideOne.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => {
        return { sideOne };
      });
    }
  };

  onSideTwoChange = (e) => {
    const sideTwo = e.target.value;
    if (!sideTwo || sideTwo.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => {
        return { sideTwo };
      });
    }
  };

  onSideThreeChange = (e) => {
    const sideThree = e.target.value;
    if (!sideThree || sideThree.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => {
        return { sideThree };
      });
    }
  };

  onSideFourChange = (e) => {
    const sideFour = e.target.value;

    if (!sideFour || sideFour.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => {
        return { sideFour };
      });
    }
  };

  onSubmit = (e) => {
    e.preventDefault();
    if (!this.state.sideOne) {
      this.setState(() => ({
        error: "Please provide at least 'Side One' ",
      }));
    } else if (
      this.state.sideOne &&
      this.state.sideTwo &&
      !this.state.sideThree
    ) {
      this.setState(() => ({
        error: "Please provide a shape with more or less than 2 sides",
      }));
    } else {
      this.setState(() => ({ error: "" }));
      this.props.onSubmit({
        sideOne: this.state.sideOne,
        sideTwo: this.state.sideTwo,
        sideThree: this.state.sideThree,
        sideFour: this.state.sideFour,
      });

      if (
        this.state.sideOne &&
        this.state.sideTwo &&
        this.state.sideThree &&
        this.state.sideFour
      ) {
        this.props.sortByQuadrilaterals();
      } else if (
        this.state.sideOne &&
        this.state.sideTwo &&
        this.state.sideThree
      ) {
        this.props.sortByTriangles();
      } else if (this.state.sideOne) {
        this.props.sortByCircles();
      }
    }
  };

  render() {
    return (
      <FormContainer>
        <Instructions
          aria-readonly="true"
          role="Instructions on how to use this website"
        >
          Please enter your the lengths of each shape of the shape within the
          fields below, consider the lengths are been measured in pixels so you
          may want to use double digits to ensure your shape is visible!
          <Instructions>
            Then I will return your shape below, sorted into it's category!
          </Instructions>
        </Instructions>
        <form onSubmit={this.onSubmit}>
          {this.state.error && <p>{this.state.error}</p>}
          <label for="Shape Side One">Side One:</label>
          <input
            autoFocus
            onChange={this.onSideOneChange}
            value={this.state.sideOne}
            placeholder="Side One"
          ></input>
          <label for="Shape Side Two">Side Two:</label>
          <input
            disabled={!this.state.sideOne ? true : false}
            onChange={this.onSideTwoChange}
            value={this.state.sideTwo}
            placeholder="Side Two"
          ></input>
          <label for="Shape Side Three">Side Three:</label>
          <input
            disabled={!this.state.sideTwo ? true : false}
            onChange={this.onSideThreeChange}
            value={this.state.sideThree}
            placeholder="Side Three"
          ></input>
          <label for="Shape Side Four">Side Four:</label>
          <input
            disabled={!this.state.sideThree ? true : false}
            onChange={this.onSideFourChange}
            value={this.state.sideFour}
            placeholder="Side Four"
          ></input>
          <button>Create new shape</button>
        </form>
      </FormContainer>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    filters: state.filters,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    sortByTriangles: () => dispatch(sortByTriangles()),
    sortByCircles: () => dispatch(sortByCircles()),
    sortByQuadrilaterals: () => dispatch(sortByQuadrilaterals()),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(ShapeForm);
