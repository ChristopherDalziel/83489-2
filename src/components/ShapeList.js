import React from "react";
import { connect } from "react-redux";
import styled from "@emotion/styled";
import ShapeListItem from "./ShapeListItem";
import sortShapes from "../selectors/shapes";

const Body = styled.div`
  margin: 10px;
  overflow: scroll;
`;

const H2 = styled.h2`
  color: #e63846;
`;

const NoShapesMessage = styled.p`
  color: #e63846;
`;

export const ShapeList = (props) => {
  return (
    <Body>
      {props.shapes.length === 0 ? (
        <NoShapesMessage>No shapes to display</NoShapesMessage>
      ) : (
        <>
          <H2>Sorted Shapes</H2>
          {props.shapes.map((shape, index) => {
            return <ShapeListItem key={index} {...shape} />;
          })}
        </>
      )}
    </Body>
  );
};

const mapStateToProps = (state) => {
  return {
    shapes: sortShapes(state.shapes, state.filters),
  };
};

export default connect(mapStateToProps)(ShapeList);
