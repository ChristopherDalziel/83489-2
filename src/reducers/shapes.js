const shapesReducerDefaultState = [];

const shapesReducer = (state = shapesReducerDefaultState, action) => {
  switch (action.type) {
    case "ADD_SHAPE":
      return [...state, action.shape];
    case "RESET_SHAPES":
      return shapesReducerDefaultState;
    default:
      return state;
  }
};

export default shapesReducer;
