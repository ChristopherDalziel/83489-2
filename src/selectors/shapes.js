const expectedSides = {
  circles: 1,
  triangles: 3,
  quadrilaterals: 4,
};

const sortShapes = (shapes, { sortBy }) => {
  return shapes.filter((shape) => {
    const sides = Object.values(shape).filter((side) => side).length;
    return sides === expectedSides[sortBy];
  });
};

export default sortShapes;
