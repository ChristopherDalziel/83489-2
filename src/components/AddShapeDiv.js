import React from "react";
import { connect } from "react-redux";
import styled from "@emotion/styled";
import ShapeForm from "./ShapeForm";
import { addShape } from "../actions/shapes";

const Button = styled.button`
  margin: 10px 0px 0px 10px;
  width: 10vw;
  @media (max-width: 815px) {
    width: 50vw;
    text-align: left;
  }
`;

export class AddShapeDiv extends React.Component {
  state = {
    menuShown: false,
  };

  onSubmit = (shape) => {
    this.props.addShape(shape);
    this.setState({ menuShown: false });
  };

  render() {
    return (
      <>
        {this.state.menuShown ? (
          <Button
            type="toggle"
            onClick={() => this.setState({ menuShown: !this.state.menuShown })}
          >
            Cancel
          </Button>
        ) : (
          <Button
            type="toggle"
            onClick={() => this.setState({ menuShown: !this.state.menuShown })}
          >
            Create your own shape
          </Button>
        )}

        {this.state.menuShown ? <ShapeForm onSubmit={this.onSubmit} /> : ""}
      </>
    );
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    addShape: (shape) => dispatch(addShape(shape)),
  };
};

export default connect(undefined, mapDispatchToProps)(AddShapeDiv);
