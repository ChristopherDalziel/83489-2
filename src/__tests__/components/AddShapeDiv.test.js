import React from "react";
import { mount } from "enzyme";
import { AddShapeDiv } from "../../components/AddShapeDiv";
import shapes from "../fixtures/shapes";

import toJSON from "enzyme-to-json";

let addShape, wrapper;

beforeEach(() => {
  addShape = jest.fn();
  wrapper = mount(<AddShapeDiv addShape={addShape} />);
});

test("Should render the AddShapeDiv correctly", () => {
  expect(toJSON(wrapper)).toMatchSnapshot();
});

// This tests failed because when it's called it is testing the connected form, if it's changed to the named import within the 'AddShapeDiv' the test passes but the app crashes.
test("Should handle onSubmit", () => {
  wrapper.find("ShapeForm").at(0).prop("onSubmit")(shapes[1]);
  expect(addShape).toHaveBeenLastCalledWith(shapes[1]);
});
