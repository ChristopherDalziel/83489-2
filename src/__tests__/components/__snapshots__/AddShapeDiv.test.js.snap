// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render the AddShapeDiv correctly 1`] = `
<AddShapeDiv
  addShape={[MockFunction]}
>
  <Styled(button)
    onClick={[Function]}
    type="toggle"
  >
    <button
      className="css-njncyf"
      onClick={[Function]}
      type="toggle"
    >
      Create your own shape
    </button>
  </Styled(button)>
</AddShapeDiv>
`;
